lib/irb/history.rb

```diff
@@ -53,26 +53,26 @@ def save_history
           raise
         end
 
-        if !File.exist?(history_file)
-          warn "Warning: The path for the HISTORY_FILE does not exist."
-        end
-
         if File.exist?(history_file) &&
            File.mtime(history_file) != @loaded_history_mtime
           history = history[@loaded_history_lines..-1] if @loaded_history_lines
           append_history = true
         end
 
-        File.open(history_file, (append_history ? 'a' : 'w'), 0o600, encoding: IRB.conf[:LC_MESSAGES]&.encoding) do |f|
-          hist = history.map{ |l| l.scrub.split("\n").join("\\\n") }
-          unless append_history
-            begin
-              hist = hist.last(num) if hist.size > num and num > 0
-            rescue RangeError # bignum too big to convert into `long'
-              # Do nothing because the bignum should be treated as infinity
+        begin
+          File.open(history_file, (append_history ? 'a' : 'w'), 0o600, encoding: IRB.conf[:LC_MESSAGES]&.encoding) do |f|
+            hist = history.map{ |l| l.scrub.split("\n").join("\\\n") }
+            unless append_history
+              begin
+                hist = hist.last(num) if hist.size > num and num > 0
+              rescue RangeError # bignum too big to convert into `long'
+                # Do nothing because the bignum should be treated as infinity
+              end
             end
+            f.puts(hist)
           end
-          f.puts(hist)
+        rescue Errno::ENOENT
```

Comment:
Instead of rescuing exception, I think early return like this is better:

```rb
pathname = Pathname.new(history_file)
unless Dir.exists?(pathname.dirname)
  warn "The folder for history file doesn't exist ....."
  return
end

File.open(history_file, (append_history ? 'a' : 'w'), 0o600, encoding: IRB.conf[:LC_MESSAGES]&.encoding) do |f|
# ...
```